airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -145,68 +173,76 @@ public synchronized void scanLibrary() {
         }
         scanning = true;
 
-        Thread thread = new Thread("MediaLibraryScanner") {
-            @Override
-            public void run() {
-                doScanLibrary();
-                playlistService.importPlaylists();
-                mediaFileDao.checkpoint();
-            }
-        };
-
-        thread.setPriority(Thread.MIN_PRIORITY);
-        thread.start();
+        ForkJoinPool pool = new ForkJoinPool(scannerParallelism, mediaScannerThreadFactory, null, true);
+        
+        CompletableFuture.runAsync(() -> doScanLibrary(), pool)
+                .thenRunAsync(() -> playlistService.importPlaylists(), pool)
+                .thenRunAsync(() -> mediaFileDao.checkpoint(), pool)
+                .thenRun(() -> pool.shutdown())
+                .thenRun(() -> scanning = false);
```

Comment:
Do in order:
- initialize the pool
- scan library (do task within the pool)
  - this will spawn more tasks within the pool that other threads will start stealing and working on
  - this is where parallelization becomes most helpful
- import playlists (do task within the pool)
- consolidate db (do task within the pool)
- shutdown the pool (outside the pool, using the common pool)
- set the scanning variable (using common pool)

This is a synchronized call, which means only 1 thread can access it at a time. thus update to scanning variable here is guaranteed to be threadsafe operation